core/corehttp/ipns_hostname.go

```diff
@@ -19,8 +19,8 @@ func IPNSHostnameOption() ServeOption {
 			defer cancel()
 
 			host := strings.SplitN(r.Host, ":", 2)[0]
-			if p, err := n.Namesys.Resolve(ctx, host); err == nil {
-				r.URL.Path = "/ipfs/" + p.String() + r.URL.Path
+			if p, err := n.Namesys.Resolve(ctx, host, 0); err == nil {
+				r.URL.Path = p.String() + r.URL.Path
```

Comment:
In order to prevent a [redirect loop](https://github.com/ipfs/go-ipfs/issues/1233#issuecomment-102618200) here please add Trimming of slashes to this line:

```diff
diff --git a/core/corehttp/ipns_hostname.go b/core/corehttp/ipns_hostname.go
index 7361001..19daefd 100644
--- a/core/corehttp/ipns_hostname.go
+++ b/core/corehttp/ipns_hostname.go
@@ -20,7 +20,7 @@ func IPNSHostnameOption() ServeOption {

                        host := strings.SplitN(r.Host, ":", 2)[0]
                        if p, err := n.Namesys.Resolve(ctx, host); err == nil {
-                               r.URL.Path = "/ipfs/" + p.String() + r.URL.Path
+                               r.URL.Path = strings.TrimRight(p.String(), "/") + r.URL.Path
                        }
                        childMux.ServeHTTP(w, r)
                })
```